File: flipperf/src/main/java/com/flipkart/flipperf/newlib/interpreter/DefaultInterpreter.java
package com.flipkart.flipperf.newlib.interpreter;

import android.support.annotation.VisibleForTesting;

import com.flipkart.flipperf.newlib.NetworkInterceptor;
import com.flipkart.flipperf.newlib.reporter.NetworkEventReporter;
import com.flipkart.flipperf.newlib.response.CountingInputStream;
import com.flipkart.flipperf.newlib.response.DefaultResponseHandler;
import com.squareup.okhttp.MediaType;
import com.squareup.okhttp.Request;
import com.squareup.okhttp.Response;
import com.squareup.okhttp.ResponseBody;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.IOException;
import java.io.InputStream;
import java.net.URL;

import okio.BufferedSource;
import okio.Okio;

public class DefaultInterpreter implements NetworkInterpreter {
    private static final String HOST_NAME = "HOST";
    private Logger logger = LoggerFactory.getLogger(DefaultInterpreter.class);
    private NetworkEventReporter mEventReporter;

    public DefaultInterpreter(NetworkEventReporter mEventReporter) {
        this.mEventReporter = mEventReporter;
    }

    @Override
    public Response interpretResponseStream(int requestId, NetworkInterceptor.TimeInfo timeInfo, Request request, Response response) throws IOException {
        final OkHttpInspectorRequest okHttpInspectorRequest = new OkHttpInspectorRequest(requestId, request.url(), request.method(), request.header("Content-Length"), request.header(HOST_NAME));
        final OkHttpInspectorResponse okHttpInspectorResponse = new OkHttpInspectorResponse(requestId, response.code(), response.header("Content-Length"), timeInfo.mStartTime, timeInfo.mEndTime);

        //if response does not have content length, using CountingInputStream to read its bytes
        if (response.header("Content-Length") == null) {
            final ResponseBody body = response.body();
            InputStream responseStream = null;
            if (body != null) {
                try {
                    responseStream = body.byteStream();
                } catch (IOException e) {
                    if (logger.isDebugEnabled()) {
                        logger.debug("Error received while reading input stream {}", e.getMessage());
                    }

                    //notify event reporter in case there is any exception while getting the input stream of response
                    mEventReporter.responseInputStreamError(okHttpInspectorRequest, okHttpInspectorResponse, e);
                    throw e;
                }
            }

            //interpreting the response stream using CountingInputStream, once the counting is done, notify the event reporter that response has been received
            responseStream = new CountingInputStream(responseStream, new DefaultResponseHandler(new DefaultResponseHandler.ResponseCallback() {
                @Override
                public void onEOF(long bytesRead) {
                    okHttpInspectorResponse.mResponseSize = String.valueOf(bytesRead);
                    mEventReporter.responseReceived(okHttpInspectorRequest, okHttpInspectorResponse);
                }
            }));

            //creating response object using the interpreted stream
            response = response.newBuilder().body(new ForwardingResponseBody(body, responseStream)).build();
        } else {
            //if response has content length, notify the event reporter that response has been received.
            mEventReporter.responseReceived(okHttpInspectorRequest, okHttpInspectorResponse);
        }
        return response;
    }

    @Override
    public void interpretError(int requestId, NetworkInterceptor.TimeInfo timeInfo, Request request, IOException e) {
        if (logger.isDebugEnabled()) {
            logger.debug("Error received while proceeding response {}", e.getMessage());
        }
        final OkHttpInspectorRequest okHttpInspectorRequest = new OkHttpInspectorRequest(requestId, request.url(), request.method(), request.header("Content-Length"), request.header(HOST_NAME));
        mEventReporter.httpExchangeError(okHttpInspectorRequest, e);
    }

    /**
     * Implementation of {@link NetworkEventReporter.InspectorRequest}
     */
    @VisibleForTesting
    public static class OkHttpInspectorRequest implements NetworkEventReporter.InspectorRequest {
        private final int mRequestId;
        private final URL mRequestUrl;
        private final String mMethodType;
        private final String mContentLength;
        private final String mHostName;

        public OkHttpInspectorRequest(int requestId, URL requestUrl, String methodType, String contentLength, String hostName) {
            this.mRequestId = requestId;
            this.mRequestUrl = requestUrl;
            this.mMethodType = methodType;
            this.mContentLength = contentLength;
            this.mHostName = hostName;
        }

        @Override
        public int requestId() {
            return mRequestId;
        }

        @Override
        public URL url() {
            return mRequestUrl;
        }

        @Override
        public String method() {
            return mMethodType;
        }

        @Override
        public String requestSize() {
            return mContentLength;
        }

        @Override
        public String hostName() {
            return mHostName;
        }
    }

    /**
     * Implementation of {@link NetworkEventReporter.InspectorResponse}
     */
    @VisibleForTesting
    public static class OkHttpInspectorResponse implements NetworkEventReporter.InspectorResponse {
        private int mRequestId;
        private long mStartTime;
        private long mEndTime;
        private int mStatusCode;
        private String mResponseSize;

        public OkHttpInspectorResponse(int requestId, int statusCode, String responseSize, long startTime, long endTime) {
            this.mRequestId = requestId;
            this.mStatusCode = statusCode;
            this.mResponseSize = responseSize;
            this.mStartTime = startTime;
            this.mEndTime = endTime;
        }

        @Override
        public int requestId() {
            return mRequestId;
        }

        @Override
        public int statusCode() {
            return mStatusCode;
        }

        @Override
        public String responseSize() {
            return mResponseSize;
        }

        @Override
        public long startTime() {
            return mStartTime;
        }

        @Override
        public long endTime() {
            return mEndTime;
        }
    }

    /**
     * Wrapper for {@link ResponseBody}
     * Will only be used in case the response does not have the content-length
     */
    @VisibleForTesting
    public static class ForwardingResponseBody extends ResponseBody {
        private final ResponseBody mBody;
        private final BufferedSource mInterceptedSource;

        public ForwardingResponseBody(ResponseBody body, InputStream interceptedStream) {
            mBody = body;
            mInterceptedSource = Okio.buffer(Okio.source(interceptedStream));
        }

        @Override
        public MediaType contentType() {
            return mBody.contentType();
        }

        @Override
        public long contentLength() throws IOException {
            return mBody.contentLength();
        }

        @Override
        public BufferedSource source() {
            return mInterceptedSource;
        }
    }
}

File: library-kotlin/src/main/java/com/flipkart/okhttpstats/kotlin/interpreter/DefaultInterpreter.kt
package com.flipkart.okhttpstats.kotlin.interpreter

import android.support.annotation.VisibleForTesting
import android.util.Log
import com.flipkart.okhttpstats.kotlin.NetworkInterceptor
import com.flipkart.okhttpstats.kotlin.reporter.NetworkEventReporter
import com.flipkart.okhttpstats.kotlin.response.CountingInputStream
import com.flipkart.okhttpstats.kotlin.response.DefaultResponseHandler
import com.flipkart.okhttpstats.kotlin.toolbox.Utils
import okhttp3.MediaType
import okhttp3.Request
import okhttp3.Response
import okhttp3.ResponseBody
import okio.BufferedSource
import okio.Okio
import java.io.IOException
import java.io.InputStream
import java.net.URL

/**
 * Default implementation of [NetworkInterpreter]
 */
class DefaultInterpreter(internal var eventReporter: NetworkEventReporter) : NetworkInterpreter {

    @Throws(IOException::class)
    override fun interpretResponseStream(requestId: Int, timeInfo: NetworkInterceptor.TimeInfo, request: Request, response: Response): Response {
        var response = response
        val responseBody = response.body()

        val okHttpInspectorRequest = OkHttpInspectorRequest(requestId, request.url().url(), request.method(), Utils.contentLength(request.headers()), request.url().host())
        val okHttpInspectorResponse = OkHttpInspectorResponse(requestId, response.code(), Utils.contentLength(response.headers()), timeInfo.startTime, timeInfo.endTime, responseBody)
        //if response does not have content length, using CountingInputStream to read its bytes
        if (response.header(CONTENT_LENGTH) == null) {
            var responseStream: InputStream? = null
            try {
                responseStream = responseBody?.byteStream()
            } catch (e: Exception) {
                if (Utils.isLoggingEnabled) {
                    Log.d("Error reading IS : ", e.message)
                }
                //notify event reporter in case there is any exception while getting the input stream of response
                eventReporter.responseInputStreamError(okHttpInspectorRequest, okHttpInspectorResponse, e)
                throw e
            }

            //interpreting the response stream using CountingInputStream, once the counting is done, notify the event reporter that response has been received
            responseStream = CountingInputStream(responseStream, DefaultResponseHandler(object : DefaultResponseHandler.ResponseCallback {
                override fun onEOF(bytesRead: Long) {
                    okHttpInspectorResponse.responseSize = bytesRead
                    eventReporter.responseReceived(okHttpInspectorRequest, okHttpInspectorResponse)
                }
            }))

            //creating response object using the interpreted stream
            response = response.newBuilder().body(ForwardingResponseBody(responseBody, responseStream)).build()
        } else {
            //if response has content length, notify the event reporter that response has been received.
            eventReporter.responseReceived(okHttpInspectorRequest, okHttpInspectorResponse)
        }

        return response
    }

    override fun interpretError(requestId: Int, timeInfo: NetworkInterceptor.TimeInfo, request: Request, e: IOException) {
        if (Utils.isLoggingEnabled) {
            Log.d("Error response: ", e.message)
        }
        val okHttpInspectorRequest = OkHttpInspectorRequest(requestId, request.url().url(), request.method(), Utils.contentLength(request.headers()), request.header(HOST_NAME))
        eventReporter.httpExchangeError(okHttpInspectorRequest, e)
    }

    /**
     * Implementation of [NetworkEventReporter.InspectorRequest]
     */
    internal class OkHttpInspectorRequest(val requestId: Int, val requestUrl: URL, val methodType: String, val contentLength: Long, val hostName: String) : NetworkEventReporter.InspectorRequest {

        override fun requestId(): Int {
            return requestId
        }

        override fun url(): URL {
            return requestUrl
        }

        override fun method(): String {
            return methodType
        }

        override fun requestSize(): Long {
            return contentLength
        }

        override fun hostName(): String {
            return hostName
        }
    }

    /**
     * Implementation of [NetworkEventReporter.InspectorResponse]
     */
    internal class OkHttpInspectorResponse(var requestId: Int, var statusCode: Int, var responseSize: Long, var startTime: Long, var endTime: Long, var responseBody: ResponseBody?) : NetworkEventReporter.InspectorResponse {

        override fun requestId(): Int {
            return requestId
        }

        override fun statusCode(): Int {
            return statusCode
        }

        override fun responseSize(): Long {
            return responseSize
        }

        override fun startTime(): Long {
            return startTime
        }

        override fun endTime(): Long {
            return endTime
        }

        override fun responseBody(): ResponseBody? {
            return responseBody
        }
    }

    /**
     * Wrapper for [ResponseBody]
     * Will only be used in case the response does not have the content-length
     */
    @VisibleForTesting
    internal class ForwardingResponseBody(val body: ResponseBody, interceptedStream: InputStream) : ResponseBody() {
        val interceptedSource: BufferedSource

        init {
            interceptedSource = Okio.buffer(Okio.source(interceptedStream))
        }

        override fun contentType(): MediaType {
            return body.contentType()
        }

        override fun contentLength(): Long {
            return body.contentLength()
        }

        override fun source(): BufferedSource {
            return interceptedSource
        }
    }

    companion object {
        private val HOST_NAME = "HOST"
        private val CONTENT_LENGTH = "Content-Length"
    }
}